web/src/message_events.js

```diff
@@ -429,6 +435,44 @@ export function update_messages(events) {
                 }
             }
 
+            // If a message was moved to the current narrow and we don't have
+            // the message cached, we need to refresh the narrow to display the message.
+            if (!changed_narrow && refreshed_current_narrow && current_filter) {
+                let is_narrowed_to_new_stream = false;
+                if (stream_changed) {
+                    const new_stream_name = sub_store.get(new_stream_id).name;
+                    is_narrowed_to_new_stream = current_filter.has_operand(
+                        "channel",
+                        new_stream_name,
+                    );
+                } else {
+                    is_narrowed_to_new_stream = current_filter.has_operand(
+                        "channel",
+                        old_stream_name,
+                    );
+                }
+
+                let is_narrowed_to_new_topic = false;
+                if (topic_edited) {
+                    is_narrowed_to_new_topic =
+                        is_narrowed_to_new_stream && current_filter.has_operand("topic", new_topic);
+                } else {
+                    is_narrowed_to_new_topic =
+                        is_narrowed_to_new_stream &&
+                        current_filter.has_operand("topic", orig_topic);
+                }
```

Comment:
Makes sense, what do you think of defining in an opposite way:

```js
    can_newly_match_moved_messages(new_stream: string, new_topic: string): boolean {
        // The moved message can be a part of lot of new narrows, especially
        // if it was moved from an unsubscribed stream for the current user.
        // So, we return true unless we are filtered to a narrow where the
        // moved message can't be a part of.
        if (this.sorted_term_types().some((term) => term.includes("dm"))) {
            return false;
        }

        if (this.has_operator("channel") && !this.has_operand("channel", new_stream)) {
            return false;
        }

        if (this.has_operator("topic") && !this.has_operand("topic", new_topic)) {
            return false;
        }

        return true;
    }
```

Reply:
I think that's worse in its handling of the absence of channel/topic operators; if you're looking at your `mentions` or `starred` view, then there's no channel/topic operators, and moving messages can't impact whether they're in that view. 

That's why I was thinking we should have a construction that looks at whether there's any operator whose true value might change because of a message being moved -- that algorithm can be precisely correct.

Reply:
I can add `mentions` and `starred` or similar views but I was thinking it is important that we show the moved message via rerendering than miss it even if we do needless rerenders for a few narrows.

Reply:
I think we just need to do the TODO I wrote out in the comment; to be clear, I believe the list I wrote to be basically complete, and I can fix it up after an extra audit if it isn't; I just didn't feel like writing out the copy-pasted logic at that time,

Reply:
Okay, updated with that code.